src/hooks/useNestedMenu.jsx

```diff
@@ -0,0 +1,143 @@
+    createReducerHelpers(menuItems)
+  );
+  const [showChildren, dispatch] = useReducer(reducer, initState);
+  const [refs, setRefs] = useState({});
+  const shouldRender = useMountDelay(menuItems, showChildren, delay);
+  const [windowDim, setWindowDim] = useState({ width: 0, height: 0 });
+
+  const debounce = useCallback(db(), []);
+
+  useEffect(() => {
+    const handleResize = () => {
+      setWindowDim({ width: window.innerWidth, height: window.innerHeight });
+    };
+
+    window.addEventListener('resize', handleResize);
+
+    () => window.removeEventListener('resize', handleResize);
+  }, []);
+
+  useEffect(() => {
+    closeNested(menuItems);
+  }, [windowDim]);
+
+  useEffect(() => {
+    for (const key in showChildren) {
+      setRefs(prevState => ({ ...prevState, [key]: createRef() }));
+    }
+    setRefs(prevState => ({ ...prevState, nav: navRef }));
+  }, []);
+
+  const closeNested = items => {
+    if (!items) return;
+    items.forEach(item => {
+      if (item.children?.length && showChildren[item.name]) {
+        dispatch(actions[`${item.name}Close`]);
+        closeNested(item.children);
+      }
+    });
+  };
+
+  const toggleNested = (item, e) => {
+    e.stopPropagation();
+    const siblings = getSiblings(item.name);
+    closeNested(siblings);
+    if (showChildren[item.name] !== undefined) {
+      dispatch(actions[`${item.name}Open`]);
+    }
+  };
+
+  const handleMenuLeave = (e, openMenu = false, delay = 500) => {
+    e.stopPropagation();
+    openMenu
+      ? debounce(0, () => {})
+      : debounce(delay * 1.5, () => closeNested(menuItems));
+  };
+
+  const openSubmenu = (name, e = null) => {
+    e?.stopPropagation();
+    if (!showChildren[name] && actions[name + 'Open']) {
+      dispatch(actions[name + 'Open']);
+    }
+  };
+
+  const getSiblings = useCallback(
+    (itemName, children = null) => {
+      let siblings = [];
+      let currentChildren = children || menuItems;
+      currentChildren.forEach(item => {
+        if (showChildren[item.name]) {
+          item.children.find(i => i.name === itemName)
+            ? (siblings = item.children)
+            : (siblings = getSiblings(itemName, item.children));
+        }
+      });
+      siblings = siblings.length
+        ? siblings.filter(({ name }) => name !== itemName)
+        : siblings;
+      return siblings;
+    },
+    [menuItems, showChildren]
+  );
+
+  const setIsItemActive = useCallback(
+    (item, parent = null) => {
+      if (!parent) return;
+      if (item.children?.length && menuItems.find(i => i === parent)) {
+        return showChildren[item.name];
+      } else {
+        const siblings = getSiblings(item.name);
+        return !siblings.some(sibling => showChildren[sibling.name]);
+      }
+    },
+    [menuItems, showChildren]
+  );
+
+  const renderMenu = (items = menuItems, isNested = false, parent = null) => {
+    return items.map(item => (
+      <LinkWrapper
```

Comment:
Everything within `map` must be put into a separate component I mentioned in the comment above. It will allow you to avoid refs array and some handlers that iterate over the whole menu structure. Think of it as a component that is only responsible for rendering its title or a nested menu if it has it. You will put all the necessary handlers to that component as well. That component will also contain its state to determine whether to show the nested menu or not.